dpcpp/solver/batch_cg_kernels.hpp.inc

```diff
@@ -0,0 +1,252 @@
+    const auto tid = item_ct1.get_local_linear_id();
+    const auto group_size = item_ct1.get_local_range().size();
+
+    // copy x from global to shared memory
+    // r = b
+    for (int iz = tid; iz < num_rows; iz += group_size) {
+        x_shared_entry[iz] = x_global_entry[iz];
+        r_shared_entry[iz] = b_global_entry[iz];
+    }
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+
+    // r = b - A*x
+    advanced_apply_kernel(static_cast<ValueType>(-1.0), mat_global_entry,
+                          x_shared_entry, static_cast<ValueType>(1.0),
+                          r_shared_entry, item_ct1);
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+
+
+    // z = precond * r
+    prec_shared.apply(num_rows, r_shared_entry, z_shared_entry, item_ct1);
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+
+    // Compute norms of rhs
+    // and rho_old = r' * z
+    if (sg_id == 0) {
+        single_rhs_compute_norm2_sg(num_rows, b_global_entry, rhs_norms,
+                                    item_ct1);
+    } else if (sg_id == 1) {
+        single_rhs_compute_conj_dot_sg(num_rows, r_shared_entry, z_shared_entry,
+                                       rho_old, item_ct1);
+    }
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+
+    // p = z
+    for (int iz = tid; iz < num_rows; iz += group_size) {
+        p_shared_entry[iz] = z_shared_entry[iz];
+    }
+}
+
+
+template <typename ValueType>
+__dpct_inline__ void update_p(
+    const int num_rows, const ValueType& rho_new_shared_entry,
+    const ValueType& rho_old_shared_entry,
+    const ValueType* const __restrict__ z_shared_entry,
+    ValueType* const __restrict__ p_shared_entry, sycl::nd_item<3> item_ct1)
+{
+    const ValueType beta = rho_new_shared_entry / rho_old_shared_entry;
+    for (int li = item_ct1.get_local_linear_id(); li < num_rows;
+         li += item_ct1.get_local_range().size()) {
+        p_shared_entry[li] = z_shared_entry[li] + beta * p_shared_entry[li];
+    }
+}
+
+template <typename ValueType>
+__dpct_inline__ void update_x_and_r(
+    const int num_rows, const ValueType rho_old_shared_entry,
+    const ValueType* const __restrict__ p_shared_entry,
+    const ValueType* const __restrict__ Ap_shared_entry,
+    ValueType& alpha_shared_entry, ValueType* const __restrict__ x_shared_entry,
+    ValueType* const __restrict__ r_shared_entry, sycl::nd_item<3> item_ct1)
+{
+    auto sg = item_ct1.get_sub_group();
+    const auto tid = item_ct1.get_local_linear_id();
+    if (sg.get_group_id() == 0) {
+        single_rhs_compute_conj_dot_sg(num_rows, p_shared_entry,
+                                       Ap_shared_entry, alpha_shared_entry,
+                                       item_ct1);
+    }
+    item_ct1.barrier(sycl::access::fence_space::global_and_local);
+    if (tid == 0) {
+        alpha_shared_entry = rho_old_shared_entry / alpha_shared_entry;
+    }
```

Comment:
don't you also need a barrier after this? In the cuda/hip this is solved by storing this in a local variable instead of shared memory.

Reply:
good catch!